### PR TITLE
[feature] ui events

### PR DESCRIPTION
**Pull request by steveruizok**
This PR updates the editor events:
- adds types to the events emitted by the app (by `app.emit`)
- removes a few events emitted by the app (e.g. `move-to-page`, `change-camera`)
- adds `onEvent` prop to the <TldrawUi> / <Tldraw> components
- call the `onEvent` when actions occur or tools are selected
- does some superficial cleanup on editor app APIs

### Release Note

- Fix layout bug in error dialog
- (ui) Add `TLEventMap` for types emitted from editor app
- (editor) Update `crash` event emitted from editor app to include error
- (editor) Update `change-history` event emitted from editor app
- (editor) Remove `change-camera` event from editor app
- (editor) Remove `move-to-page` event from editor app
- (ui) Add `onEvent` prop and events to <Tldraw> / <TldrawUi>
- (editor) Replace `app.openMenus` plain Set with computed value
- (editor) Add `addOpenMenu` method
- (editor) Add `removeOpenMenu` method
- (editor) Add `setFocusMode` method 
- (editor) Add `setToolLocked` method  
- (editor) Add `setSnapMode` method 
- (editor) Add `isSnapMode` method 
- (editor) Update `setGridMode` method return type to editor app
- (editor) Update `setReadOnly` method return type to editor app
- (editor) Update `setPenMode` method return type to editor app
- (editor) Update `selectNone` method return type to editor app
- (editor) Rename `backToContent` to `zoomToContent`
- (editor) Remove `TLReorderOperation` type

### CODE REVIEW

File: packages/editor/src/lib/app/managers/HistoryManager.ts [hunk 1]
```diff
@@ -103,7 +106,7 @@ export class HistoryManager<CTX extends { emit: (name: string) => void }> {
 					this._redos.set(stack())
 				}
 
-				this.ctx.emit('change-history')
+				this.ctx.emit('change-history', { reason: 'push' })
```
Comment (steveruizok) [inline packages/editor/src/lib/app/managers/HistoryManager.ts:109]: Adding reasons to our change-history events

File: packages/editor/src/lib/hooks/useTransform.ts [hunk 2]
```diff
@@ -14,7 +14,7 @@ export function useTransform(
 
 		if (scale !== undefined) {
 			if (rotate !== undefined) {
-				elm.style.transform = `translate(${x}px, ${y}px) scale(${scale}) rotate(${rotate}deg)`
+				elm.style.transform = `translate(${x}px, ${y}px) scale(${scale}) rotate(${rotate}rad)`
```
Comment (steveruizok) [inline packages/editor/src/lib/hooks/useTransform.ts:17]: Need to roll this fix into a different PR

File: packages/editor/src/lib/components/DefaultErrorFallback.tsx [hunk 3]
```diff
@@ -40,11 +40,11 @@ export const DefaultErrorFallback: TLErrorFallback = ({ error, app }) => {
 		},
 		[app]
 	)
-	const [isDarkMode, setIsDarkMode] = useState<null | boolean>(null)
```
Comment (steveruizok) [inline packages/editor/src/lib/components/DefaultErrorFallback.tsx:43]: There is some unrelated consistency changes here. For our editor, we had both `isBoolean / setIsBoolean` and `isBoolean / setBoolean`, and I consolidated on `isBoolean / setBoolean` (for example, `isDarkMode` / `setDarkMode`). I've here also rolled this out to other usage in the app.

File: packages/editor/src/lib/test/commands/stackShapes.test.ts [hunk 4]
```diff
@@ -51,7 +51,7 @@ describe('distributeShapes command', () => {
 	describe('when less than three shapes are selected', () => {
 		it('does nothing', () => {
 			const fn = jest.fn()
-			app.on('update_node', fn)
```
Comment (steveruizok) [inline packages/editor/src/lib/test/commands/stackShapes.test.ts:54]: Updated reference to a very, very old events API.

File: .prettierignore [hunk 5]
```diff
@@ -8,4 +8,6 @@
 **/.tsbuild*
 **/.next/*
 *.mdx
-**/_archive/*
\ No newline at end of file
+**/_archive/*
+**/*.json
+**/*.js
```
Comment (SomeHats) [inline .prettierignore:13]: not sure if this is from steves or jamies changes but we generally do want to include .json & .js files in prettier! we've had issues before with unrelated formatting changes getting into diffs with these files because some team members don't have prettier configured on save. Could we keep these included? Are there specific files causing problems?

File: packages/editor/src/lib/hooks/usePattern.tsx [hunk 6]
```diff
@@ -88,7 +88,7 @@ const getDefaultPatterns = () => {
 export const usePattern = () => {
 	const app = useApp()
 	const dpr = app.devicePixelRatio
-	const [isReady, setIsReady] = useState(false)
+	const [isReady, setReady] = useState(false)
```
Comment (orangemug) [inline packages/editor/src/lib/hooks/usePattern.tsx:91]: The norm is `something` / `setSomething` why the change?

**Comment by orangemug** [comment]
I don't think this is the right approach (if I understand correctly). The previous branch used the core data structures to avoid losing changes, and making you think about events every time you make a change. For example

https://github.com/tldraw/tldraw/blob/2d8da6a1b037bfe88a261a94104b987b4960dda9/packages/editor/src/lib/app/App.ts#L1479-L1483

The above means that if I call `this.updateUserDocumentSettings({ isDarkMode }, true)` by itself it'll miss the event entirely. An example, if I wanted made a method that synced settings across tabs. I would likely just call `updateUserDocumentSettings` updating all the settings in one shot. With these changes I'd have to write a separate event handler for each way I want to use the core data.

In my experience leads to missed data / events. If you were using this for analytics for example it could be months of incorrect data / decisions before the team discover the issue, I've personally seen this may times.

I'd prefer to derive the data from the store. In the original commits I was using the `onAfterChange` for this but I also think we could just use the `computed` from signia (CC @ds300).

Either way I think we want a single change that produces side effects (react-ey) rather than putting the side-effects on various functions around the codebase.

**Comment by steveruizok** [comment]
My main concern there is that there are some intents, such as duplicating a page, that can't be seen by examining the changes alone. There are also changes that, if observed, could have be triggered by many changes: for example "create page" being caused by duplicating a page, creating a new page, moving shapes to a new page, undoing a delete page, redoing a create page, etc. There's also the case where actions from other users might get reported as having come from the user, for example when a page is added to the store by a collaborator. 😮‍💨

At this point, I think it would be better to catch and report changes based on the intent we want to capture, rather than specifically what happened in the store. tldraw really isn't an event-driven application, so if a developer really wanted to know every time a shape appeared on the canvas (for any reason), I'd rather point them to the `store.listen` method. Maybe we can prefix the events here to be more explicit: `user-create-shapes`, `user-duplicate-page`, etc?

I'd earlier considered moving certain events (like toggling dark mode) out to the UI actions in `useActions`, so that the event is emitted when the action runs rather than when the API method runs. This wasn't a full solution, however, as we'd also need to emit changes like `create-shapes` from the editor alone.

**Comment by steveruizok** [comment]
Ok, I think this is good discussion. Looking back at the events, there seem to be a few different categories:

### Changes

Things that happen as the result of some change to the store. For example, the `currentPageId` may change for many reasons, such as when deleting the current page, duplicating a page, moving shapes to a page, or having someone else delete your current page. Likewise, shape records may be created or removed for many reasons.

- `create-shapes`
- `delete-shapes`
- `create-page`
- `delete-page`
- `change-camera`
- `change-setting`

These are all events that **could be seen directly from the store**, and so could be emitted automatically via reactors or via the `onAfterCreate` / `onAfterDelete` / `onAfterUpdate` calls as @orangemug did [here](https://github.com/tldraw/tldraw/compare/lite...feature/ui-events#diff-1e605f63c52ca3f75dca38ca233130db765f5a3d6944a4359a2af6e4ee664bf2R260).

### Editor Lifecycle

Events that refer to moments within the editor's life cycle:
- `mount`
- `crash`
- `change`
- `update`
- `change-history`
- `event`
- `max-shapes`
- `stop-camera-animation`
- `change-path`
- `change-history`
- `mark-history`

### Actions

Higher-level method calls, such as `App.zoomIn` or `App.duplicateShapes`, which make lower changes to some records. An example being `App.zoomIn`, which makes a change to the `camera` record, or `App.duplicateShapes`, which calls other methods (`App.createShapes`). These may be trigged by calling the methods or by interacting with our UI (which also calls the methods).

- `group-shapes`
- `ungroup-shapes`
- `duplicate-page`
- `duplicate-page`
- `pack-shapes`
- `stretch-shapes`
- `reorder-shapes`
- `flip-shapes`
- `distribute-shapes`
- `align-shapes`
- `set-prop`
- `set-tool`
- `rotate-shapes`
- `nudge-shapes`
- `select-all`
- `select-none`
- `zoom-in`
- `zoom-out`
- `zoom-to-selection`
- `reset-zoom`
- `zoom-into-view`
- `zoom-to-content`

### Pure-ui Events

Events that really only 

- `open-menu`
- `close-menu`
- `create-new-project`
- `save-project-to-file`
- `open-file`
- `print`
- `cut`
- `copy`
- `paste`
- `share`
- `copy-share-link`
etc.

I think the move here would be to skip the **change** events (or rather, to subscribe to them using reactors). I'd like to have / share analytics for when users use tools to create shapes, however we may be able to derive shape use by looking at other data.

**Comment by steveruizok** [comment]
I'll merge in some changes when I get to the office that brought this PR in line with my comment above.

**Comment by orangemug** [comment]
> > tldraw really isn't an event-driven application
> 
> ...If we want to add events, I think we should consider that and make that a possible approach to develop an app against.

Sorry @steveruizok I guess what I meant by this was our intention here is to get analytics into the app. We seem to have diverged into adding general purpose events.

I think that's a different task, and would probably require a little more research. So I think to do that well we should look at existing libraries like [maplibre-gl-js events](https://maplibre.org/maplibre-gl-js-docs/api/map/#map-events) and leaflet and other also other drawing apps. I think a good API here is something that feels natural, something you can almost guess because you've used similar libraries before with a camera / drawing tools etc...

If we just use a callback for now we can mark the whole API as `@internal` and leave this decision for another day.

Otherwise we're going to have to mark particular events from the event emitter as `@public`/`@internal` which seems to add more confusion to our API.

We could also research and general purpose events API. But I'd like to have a reason for adding that and some aims/usecases first.

File: .prettierignore (hunk 5 above)
Comment (steveruizok) [reply .prettierignore:13]: Fixing!

File: packages/editor/src/lib/hooks/usePattern.tsx (hunk 6 above)
Comment (steveruizok) [reply packages/editor/src/lib/hooks/usePattern.tsx]: Fixed!

File: .ignore [hunk 7]
```diff
@@ -0,0 +1,21 @@
+dist
```
Comment (steveruizok) [inline .ignore:1]: This prevents these files from getting into our search results. Could be part of a different PR. Doesn't effect anything at runtime.

File: packages/editor/src/lib/app/types/emit-types.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
+import { TLPageId, TLRecord } from '@tldraw/tlschema'
+import { TLChange } from '../App'
+import { TLEventInfo } from './event-types'
+
+/** @public */
+export interface TLEventMap {
```
Comment (steveruizok) [inline packages/editor/src/lib/app/types/emit-types.ts:6]: These are the types that we allow to be emitted by `app.emit`. They used to be untyped, now they're typed!

File: packages/editor/src/lib/test/App.test.ts [hunk 9]
```diff
@@ -202,7 +202,7 @@ describe("App's default tool", () => {
 	})
 	it('Is hand for readonly mode', () => {
 		app = new TestApp()
-		app.setReadOnly(true)
+		app.setIsReadOnly(true)
```
Comment (steveruizok) [inline packages/editor/src/lib/test/App.test.ts:205]: Unifying this in the other direction now, `isBoolean` and `setIsBoolean`.

File: packages/editor/src/lib/test/tools/TLArrowTool.test.ts [hunk 10]
```diff
@@ -111,7 +111,7 @@ describe('When dragging the arrow', () => {
 	})
 
 	it('returns to arrow.idle, keeping shape, on pointer up when tool lock is active', () => {
-		app.updateInstanceState({ isToolLocked: true })
+		app.setIsToolLocked(true)
```
Comment (steveruizok) [inline packages/editor/src/lib/test/tools/TLArrowTool.test.ts:114]: Using the new methods.

File: packages/ui/src/lib/components/MoveToPageMenu.tsx [hunk 11]
```diff
@@ -10,6 +11,7 @@ export const MoveToPageMenu = track(function MoveToPageMenu() {
 	const pages = app.pages
 	const currentPageId = app.currentPageId
 	const msg = useTranslation()
+	const { addToast } = useToasts()
```
Comment (orangemug) [inline packages/ui/src/lib/components/MoveToPageMenu.tsx:14]: Why are we adding toasts in this PR? It's also missing translations.

File: packages/ui/src/lib/hooks/useEventsProvider.tsx [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+// 	const out: Record<string, any> = {}
+// 	for (const [key, value] of Object.entries(obj)) {
+// 		if (filterValue(key) !== REDACTED_VALUE) {
+// 			out[key] = fn(value)
+// 		}
+// 	}
+// 	return out
+// }
+
+// export function filterSensitiveData(raw: any): any {
```
Comment (orangemug) [inline packages/ui/src/lib/hooks/useEventsProvider.tsx:149]: This stuff is quite important as it makes sure we don't track uses by mistake.

File: packages/ui/src/lib/hooks/useMenuIsOpen.ts [hunk 13]
```diff
@@ -1,27 +1,31 @@
...
-				app.openMenus.add(id)
-			} else {
-				app.openMenus.delete(id)
-				app.openMenus.forEach((menuId) => {
-					if (menuId.startsWith(id)) {
-						app.openMenus.delete(menuId)
-					}
-				})
-			}
+			app.batch(() => {
```
Comment (orangemug) [inline packages/ui/src/lib/hooks/useMenuIsOpen.ts:14]: This looks unrelated, different PR please.

File: public-yarn.lock [hunk 14]
```diff
@@ -9463,6 +9470,7 @@ __metadata:
     "@babel/plugin-proposal-decorators": ^7.21.0
     "@tldraw/assets": "workspace:*"
     "@tldraw/tldraw": "workspace:*"
+    "@vercel/analytics": ^1.0.1
```
Comment (orangemug) [inline public-yarn.lock:9473]: Why are we adding `@vercel/analytics` to the open source repo?

File: packages/ui/src/lib/hooks/useClipboardEvents.ts [hunk 15]
```diff
@@ -969,22 +970,27 @@ const handleNativeClipboardPaste = async (
...
 export function useMenuClipboardEvents() {
 	const app = useApp()
+	const event = useEvents()
 
 	const copy = useCallback(
 		function onCopy() {
 			if (app.selectedIds.length === 0) return
+
 			handleMenuCopy(app)
+			event('menu', 'copy')
```
Comment (orangemug) [inline packages/ui/src/lib/hooks/useClipboardEvents.ts:980]: `event(...)` isn't very clear. The `event` is usually reserved for events. A better name would be `trackEvent(...)` or something like that.

File: apps/examples/package.json [hunk 16]
```diff
@@ -31,18 +31,17 @@
 		"build": "vite build",
 		"lint": "yarn run -T tsx ../../scripts/lint.ts"
 	},
-	"devDependencies": {
-		"@babel/plugin-proposal-decorators": "^7.21.0",
-		"lazyrepo": "0.0.0-alpha.26",
-		"vite": "^4.3.4"
-	},
 	"dependencies": {
+		"@babel/plugin-proposal-decorators": "^7.21.0",
```
Comment (orangemug) [inline apps/examples/package.json:35]: These seemed correct previously, why are they now `dependencies` rather than `devDependencies`?

File: packages/editor/api-report.md [hunk 17]
```diff
@@ -355,7 +366,7 @@ export class App extends EventEmitter {
         description: string;
     }>;
     get onlySelectedShape(): TLBaseShape<any, any> | null;
-    openMenus: Set<string>;
```
Comment (orangemug) [inline packages/editor/api-report.md:358]: This makes this PR a `major` breaking change, you currently have it marked as a `patch`

File: packages/editor/src/lib/app/managers/HistoryManager.ts [hunk 18]
```diff
@@ -176,13 +179,19 @@ export class HistoryManager<CTX extends { emit: (name: string) => void }> {
 					redos = redos.push(mark)
 				}
 				if (mark.id === toMark) {
-					this.ctx.emit('change-history')
+					this.ctx.emit(
+						'change-history',
+						pushToRedoStack ? { reason: 'undo' } : { reason: 'bail', markId: toMark }
```
Comment (orangemug) [inline packages/editor/src/lib/app/managers/HistoryManager.ts:184]: What does `bail` mean?

File: packages/editor/src/lib/app/shapeutils/TLVideoUtil/TLVideoUtil.tsx [hunk 19]
```diff
@@ -152,7 +152,7 @@ const TLVideoUtilComponent = track(function TLVideoUtilComponent(props: {
 				video.pause()
 			}
 
-			setIsLoaded(true)
+			setLoaded(true)
```
Comment (orangemug) [inline packages/editor/src/lib/app/shapeutils/TLVideoUtil/TLVideoUtil.tsx:155]: Should be `setIsLoaded(...)`

**Comment by orangemug** [comment]
I'm not sure where this PR is at the moment. We seem to be rewriting the API, @SomeHats is the API person by the sounds of things

_A few comments from me_: If we are writing the API `setIsFocusMode(...)` should probably be `setFocusMode(...)` this `is` comment from an earlier PR comment was refering to the norms of `[isFoo, setFoo] = useState(...)` usage.

It's more normal to see `app.setFocusMode(boolean)` in the wild. Don't think I've ever seen something set**Is**Someting in an API (maybe others have).

I'd probably just design the API we want, and do that in a single PR. Else we'll have to _major_ update way too often for library consumers. I think the main API should rarely change, it's also probably discussing and designing in a team given together we'll probably have a good knowledge of existing APIs and how we can fit into the norm (RFC?)

**Comment by orangemug** [comment]
Also "release note" (in PR description) doesn't match the  changes we should be outlining the API changes in this section.

File: packages/editor/src/lib/app/managers/HistoryManager.ts (hunk 18 above)
Comment (steveruizok) [reply packages/editor/src/lib/app/managers/HistoryManager.ts:184]: Bail is when we undo back to the most recent mark without adding anything to the redo stack. Used most often when pressing `escape` during an interaction.

File: apps/examples/package.json (hunk 16 above)
Comment (steveruizok) [reply apps/examples/package.json:35]: The dependencies / devDependencies here really don't really matter, as this is not published to npm

File: packages/ui/src/lib/hooks/useClipboardEvents.ts (hunk 15 above)
Comment (steveruizok) [reply packages/ui/src/lib/hooks/useClipboardEvents.ts:980]: `trackEvent` is fine for me, or we could use `onEvent` as that's also the name of the handler?

File: public-yarn.lock (hunk 14 above)
Comment (steveruizok) [reply public-yarn.lock:9473]: Good catch; at one point I was experimenting with that as well, I'll yank it.

File: packages/ui/src/lib/hooks/usePrint.ts [hunk 20]
```diff
@@ -130,9 +130,9 @@ export function usePrint() {
 
 			const afterPrintHandler = () => {
 				// TODO: This is kind of lazy at the moment. I guess we need an event for 'something-happens-on-canvas'
-				app.once('change-camera', () => {
+				setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (steveruizok) [reply packages/ui/src/lib/hooks/usePrint.ts:133]: This PR removes 'change-camera' as an event. It was only used here; we could instead change the event to `change-history` rather than use a timeout.

File: packages/ui/src/lib/components/MoveToPageMenu.tsx (hunk 11 above)
Comment (steveruizok) [reply packages/ui/src/lib/components/MoveToPageMenu.tsx:14]: This PR removes the `move-to-page` event and moves that logic from out of `useAppEvents` to the callback here.

File: packages/ui/src/lib/hooks/useEventsProvider.tsx (hunk 12 above)
Comment (steveruizok) [reply packages/ui/src/lib/hooks/useEventsProvider.tsx:149]: This seemed like a part of your branch where the events that were triggered as the result of changes to the store. We're no longer hooking into direct record changes, so it's probably more useful if we end up hooking into those sorts of changes externally (e.g. from the nextjs app).
